sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/StreamOffsetRangeTrackerTest.java

```diff
@@ -0,0 +1,455 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.gcp.bigquery;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+import com.google.cloud.bigquery.storage.v1alpha1.Storage.Stream;
+import com.google.cloud.bigquery.storage.v1alpha1.Storage.StreamPosition;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** Tests for {@link StreamOffsetRangeTracker}. */
+@RunWith(JUnit4.class)
+public class StreamOffsetRangeTrackerTest {
+  @Rule public final ExpectedException expected = ExpectedException.none();
+
+  private static final Stream defaultStream = Stream.newBuilder().setName("defaultStream").build();
+
+  @Test
+  public void testGetStartPosition() throws Exception {
+    StreamOffsetRangeTracker tracker = new StreamOffsetRangeTracker(defaultStream, 100, 200, false);
+    StreamPosition startPosition =
+        StreamPosition.newBuilder().setStream(defaultStream).setOffset(100).build();
+    assertEquals(startPosition, tracker.getStartPosition());
+    assertEquals(100, tracker.getStartOffset());
+  }
+
+  @Test
+  public void testGetStopPosition() throws Exception {
+    StreamOffsetRangeTracker tracker = new StreamOffsetRangeTracker(defaultStream, 100, 200, false);
+    StreamPosition stopPosition =
+        StreamPosition.newBuilder().setStream(defaultStream).setOffset(200).build();
+    assertEquals(stopPosition, tracker.getStopPosition());
+    assertEquals(200, tracker.getStopOffset());
+  }
+
+  @Test
+  public void testTryReturnRecordFailsWithWrongStreamContext() throws Exception {
+    StreamOffsetRangeTracker tracker = new StreamOffsetRangeTracker(defaultStream, 100, 200, false);
+    StreamPosition recordStart =
+        StreamPosition.newBuilder()
+            .setStream(Stream.newBuilder().setName("newStream").build())
+            .setOffset(100)
+            .build();
+    expected.expect(IllegalStateException.class);
```

Comment:
Since `tryReturnRecordAt` can throw `IllegalStateException` in multiple situations, it might be good to add a check for a full or partial message text match with [ExpectedException.html#expectMessage](https://junit.org/junit4/javadoc/4.12/org/junit/rules/ExpectedException.html#expectMessage(java.lang.String)).

That way we can ensure that the failure is for the test case the test method is targeting.

Similar for other tests in this file.

Reply:
I didn't add coverage for the exception messages here since I was following the pattern in OffsetRangeTrackerTest.java. There's no harm in adding these checks to the tests for both files, though.